README.md
# TSS - Threshold Secret Sharing

[![Gem Version](https://badge.fury.io/rb/tss.svg)](https://badge.fury.io/rb/tss)
[![Dependency Status](https://gemnasium.com/badges/github.com/grempe/tss-rb.svg)](https://gemnasium.com/github.com/grempe/tss-rb)
[![Build Status](https://travis-ci.org/grempe/tss-rb.svg?branch=master)](https://travis-ci.org/grempe/tss-rb)
[![Coverage Status](https://coveralls.io/repos/github/grempe/tss-rb/badge.svg?branch=master)](https://coveralls.io/github/grempe/tss-rb?branch=master)
[![Code Climate](https://codeclimate.com/github/grempe/tss-rb/badges/gpa.svg)](https://codeclimate.com/github/grempe/tss-rb)

Ruby Docs : [http://www.rubydoc.info/gems/tss](http://www.rubydoc.info/gems/tss)


## WARNING : BETA CODE

This code is new and has not yet been tested in production. Use at your own risk.
The share format and interface should be fairly stable now but should not be
considered fully stable until v1.0.0 is released.

## About TSS

This Ruby gem implements Threshold Secret Sharing, as specified in
the Network Working Group Internet-Draft submitted by D. McGrew
([draft-mcgrew-tss-03.txt](http://tools.ietf.org/html/draft-mcgrew-tss-03)).

Threshold Secret Sharing (TSS) provides a way to generate `N` shares
from a value, so that any `M` of those shares can be used to
reconstruct the original value, but any `M-1` shares provide no
information about that value. This method can provide shared access
control on key material and other secrets that must be strongly
protected.

This threshold secret sharing method is based on polynomial interpolation in
GF(256) and also provides a robust format for the storage and transmission
of shares.

The sharing format is Robust Threshold Secret Sharing (RTSS) as described
in the Internet-Draft. RTSS is a binary data format and a method for
ensuring that any secrets recovered are identical to the secret that was
originally shared.

This implementation supports RTSS digest types for `NONE`, `SHA1`, and
`SHA256`. `SHA256` is the recommended digest. In the RTSS scheme a digest of
the original secret is concatenated with the secret itself prior to the splitting
of the secret into shares. Later, this digest is compared with any secret recovered
by recombining shares. If the hash of the recovered secret does not match the
original hash stored in the shares the secret will not be returned. The verifier
hash for the secret is not available to shareholders prior to recombining shares.

The specification also addresses the optional implementation of a `MAGIC_NUMBER` and
advanced error correction schemes. These extras are not currently implemented.

## TL;DR

No time for docs? Here is how to get going in a minute, using the
command line or Ruby, with the default `3 out of 5` secret sharing.

### Command Line Interface (CLI)

```sh
$ echo 'secret unicode characters ½ ♥ 💩' | tss split -O /tmp/shares.txt
$ tss combine -I /tmp/shares.txt

RECOVERED SECRET METADATA
*************************
hash : 6d1bc4242998bc170d35d374ebdf39295b271db900de25d249563a4622f113e0
hash_alg : SHA256
identifier : 6b02cfcfc24a2b50
process_time : 0.55ms
threshold : 3
secret :
secret unicode characters ½ ♥ 💩
```

### Ruby

```ruby
$ bundle exec bin/console
> require 'tss'
=> false
> shares = TSS.split(secret: 'my deep dark secret')
=> ["tss~v1~506168fade769236~3~NTA2MTY4ZmFkZTc2OTIzNgIDAEEBIM39jPGXFz4zKCObWgp2zQmCuUx92-VUf48FWQFqnLF3bw6VtVjYK7JRfESZIREdzhWcQuTQVuGKazxWRK27Tg==",
 "tss~v1~506168fade769236~3~NTA2MTY4ZmFkZTc2OTIzNgIDAEEC1E8YyDhp5NeZvF6vHeUT6HoiU0AgoF69jyjNRbtcIi1YoymJTau1rJP-3nQXOofaB2LgnAhJpbB8vrID9WTKgQ==",
 "tss~v1~506168fade769236~3~NTA2MTY4ZmFkZTc2OTIzNgIDAEEDmfvFIKybg8nO9Q9fZ5wARgHFngFQdrbk_arFEbc7s5HedTjzkoZYLlV8xnywt4AVAHAO0nSLY3C7iJPifH5VYA==",
 "tss~v1~506168fade769236~3~NTA2MTY4ZmFkZTc2OTIzNgIDAEEEiBoCoHycMNTS5yQEU8_Sc7eVlIZOJP1-ka_7MPTlC1p2DyNdGvQxy3pBFZyH8WXAY9ICBxiZRDCJisFrebviAg==",
 "tss~v1~506168fade769236~3~NTA2MTY4ZmFkZTc2OTIzNgIDAEEFxa7fSOhuV8qFrnX0KbbB3cxyWcc-8hUn4y3zZPiCmubw2TInxdncSbzDDZQgfGIPZMDsSWRbgvBOvOCK8KF94w=="]
> secret = TSS.combine(shares: shares)
=> {:hash=>"f1b91fef6a7535a974d3644c3eac16d2c907720c981290214d5d1db7cdb724af",
 :hash_alg=>"SHA256",
 :identifier=>"506168fade769236",
 :process_time=>0.94,
 :secret=>"my deep dark secret",
 :threshold=>3}
> puts secret[:secret]
my deep dark secret
=> nil
```

## Is it any good?

While this implementation has not yet had a formal security review, the cryptographic
underpinnings were carefully specified in an IETF draft document authored by a
noted cryptographer. I have reached out to individuals respected in the field
for their work in implementing cryptographic solutions to help review this code.

> I've read draft-mcgrew-tss-03 and then took a look at your code.
> Impressive! Nice docs, clean easy-to-read code. I'd use constant-time
> comparison for hashes [[resolved : 254ecab](https://github.com/grempe/tss-rb/commit/254ecab24a338872a5b05c7446213ef1ddabf4cb)],
> but apart from that I have nothing to add. Good job!
>
> -- Dmitry Chestnykh ([@dchest](https://github.com/dchest))
>
> [v0.1.0 : 4/13/2016]

All that being said, if your threat model includes a **N**ation **S**tate **A**ctor
the security of this particular code should probably not be your primary concern.

## Suggestions for Use

* Don't split large texts. Instead, split the much smaller encryption
keys that protect encrypted large texts. Supply the encrypted
files and the shares separately to recipients. Threshold secret sharing can be
very slow at splitting and recombining very large bodies of text, especially
when combined with a large number of shares. Every byte of the secret must
be processed `num_shares` times.

* Don't treat shares like encrypted data, but instead like the encryption keys
that unlock the data. Shares are keys, and need to be protected as such. There is
nothing to slow down an attacker if they have access to enough shares.

* If you send keys by email, or some other insecure channel, then your email
provider, or any entity with access to their data, now also has the keys to
your data. They just need to collect enough keys to meet the threshold. Sending
enough shares to recreate a secret through any single provider offers no
more security than sending the key itself.

* Use public key cryptography to encrypt secret shares with the public key of
each individual recipient. This can protect the share data from unwanted use while
in transit or at rest. OpenPGP would be one such tool for encrypting shares to
send safely.

* Put careful thought into how you want to distribute shares. It often makes
sense to give individuals more than one share.

## Documentation

There is pretty extensive inline documentation. You can view the latest
auto-generated docs at [http://www.rubydoc.info/gems/tss](http://www.rubydoc.info/gems/tss)

## Supported Ruby Versions

TSS is continuously integration tested on a number of Ruby versions. See the file
`.travis.yml` in the root of this repository for the currently tested versions.

## Installation

Add this line to your application's `Gemfile`:

```ruby
gem 'tss', '~> 0.4'
```

And then execute:
```sh
$ bundle
```

Or install it yourself as:

```sh
$ gem install tss
```

### Installation Security : Signed Ruby Gem

The TSS gem is cryptographically signed. To be sure the gem you install hasn’t
been tampered with you can install it using the following method:

Add my public key (if you haven’t already) as a trusted certificate

```
# Caveat: Gem certificates are trusted globally, such that adding a
# cert.pem for one gem automatically trusts all gems signed by that cert.
gem cert --add <(curl -Ls https://raw.github.com/grempe/tss-rb/master/certs/gem-public_cert_grempe_2026.pem)
```

To install, it is possible to specify either `HighSecurity` or `MediumSecurity`
mode. Since the `tss` gem depends on one or more gems that are not cryptographically
signed you will likely need to use `MediumSecurity`. You should receive a warning
if any signed gem does not match its signature.

```
# All dependent gems must be signed and verified.
gem install tss -P HighSecurity
```

```
# All signed dependent gems must be verified.
gem install tss -P MediumSecurity
```

```
# Same as above, except Bundler only recognizes
# the long --trust-policy flag, not the short -P
bundle --trust-policy MediumSecurity
```

You can [learn more about security and signed Ruby Gems](http://guides.rubygems.org/security/).

### Installation Security : Signed Git Commits

Most, if not all, of the commits and tags to the repository for this code are
signed with my PGP/GPG code signing key. I have uploaded my code signing public
keys to GitHub and you can now verify those signatures with the GitHub UI.
See [this list of commits](https://github.com/grempe/tss-rb/commits/master)
and look for the `Verified` tag next to each commit. You can click on that tag
for additional information.

You can also clone the repository and verify the signatures locally using your
own GnuPG installation. You can find my certificates and read about how to conduct
this verification at [https://www.rempe.us/keys/](https://www.rempe.us/keys/).

## Command Line Interface (CLI)

When you install the gem a simple `tss` command-line interface (CLI)
is also installed and should be available on your `$PATH`.

The CLI is a user interface for splitting and combining secrets. You can
see the full set of options available with `tss help`, `tss help split`,
or `tss help combine`.

### CLI Secret Splitting

A secret to be split can be provided using one of three
different input methods; `STDIN`, a path to a file, or when prompted
interactively. In all cases the secret should be `UTF-8` or
`US-ASCII` encoded text and be no larger than 65,535 Bytes
(including header and hash verification bytes).

Each method for entering a secret may have pros and cons from a security
perspective. You need to understand what threat model you want to protect
against in order to choose wisely. Here are a few examples for how to
split a secret.

**Example : `STDIN`**

Can read the secret from `STDIN` and write the split shares to a file. Be
cautioned that this method may leave command history which may contain your
secret.

```text
echo 'a secret' | tss split -O /tmp/shares.txt
```

**Example : `--input-file`**

Can read the secret from a file and write the split shares to `STDOUT`. Be
cautioned that storing the secret on a filesystem may expose you to certain
attacks since it can be hard to fully erase files once written.

```text
$ cat /tmp/secret.txt
my secret
$ tss split -I /tmp/secret.txt
tss~v1~8b66eb89ee25a46c~3~OGI2NmViODllZTI1YTQ2YwIDACsBqgQvwFhKboLMAmYwF5_CvaUwM8pmqUipbMRzakdbP53WJa-E6tf2nl2M
tss~v1~8b66eb89ee25a46c~3~OGI2NmViODllZTI1YTQ2YwIDACsCk1EU-HwvC6pjvQ5wDas221Qs4A7TtJNpi-Su8hxOqrsCSQ8t11aiQUpm
tss~v1~8b66eb89ee25a46c~3~OGI2NmViODllZTI1YTQ2YwIDACsDVCwbS0EGF03btYwqqVuk7S0z-nSinHtPpsFaFm5dys85j3JlK-yCVNyP
tss~v1~8b66eb89ee25a46c~3~OGI2NmViODllZTI1YTQ2YwIDACsEPUzgi8CDn4ix1dLQTQDj2yTdeeQcZAvGuMyQ_H7EepN6WO_KB1DuqTxm
tss~v1~8b66eb89ee25a46c~3~OGI2NmViODllZTI1YTQ2YwIDACsF-jHvOP2qg28J3VCK6fBx7V3CY55tTOPglelkGAzXGudBnpKC--rOvKqP
```

**Example : `interactive`**

Can read the secret interactively from a terminal and write the split shares
to `STDOUT`. This method is more secure since the secret will not be stored
in the command shell history or in a file on a filesystem that is hard to
erase securely. It does not protect you from simple keylogger attacks though.

```text
$ tss split
Enter your secret, enter a dot (.) on a line by itself to finish :
secret >  the vault password is:
secret >  V0ulT!
secret >  .
tss~v1~1f8532a9d185efc4~3~MWY4NTMyYTlkMTg1ZWZjNAIDAD8BJkMvkHUGPQk1HQr_jVj9JkhavJVkdeFDEtRAf8O6vjksCjMMjiyuUL0bOD7vDnmdrzHK8QVymXU5vnjAREs=
tss~v1~1f8532a9d185efc4~3~MWY4NTMyYTlkMTg1ZWZjNAIDAD8CJpiKCEtSfmugr_gbvEIBs2M2sZfHxuZsBuPpB35OGeyWmGZCyJCNu0W2z3abkV1Q8uTPvT75YYRpw6BcdvQ=
tss~v1~1f8532a9d185efc4~3~MWY4NTMyYTlkMTg1ZWZjNAIDAD8DdLPAuEg1Ng7hkoKFQmmL-lkILWvQiQ15JELFLJz-PvdZucDCqS-X6QAIbwWE3u2XrTgsH1kmIvpMeZhzfPk=
tss~v1~1f8532a9d185efc4~3~MWY4NTMyYTlkMTg1ZWZjNAIDAD8EJT1XJ-DeOXuE3JD6VpZk79945_peSk1ij1Mk5ql-l2jkv85yGBbyH0VIBmJp2WsEMmESZqTsQ9kKB-0Ljq8=
tss~v1~1f8532a9d185efc4~3~MWY4NTMyYTlkMTg1ZWZjNAIDAD8FdxYdl-O5cR7F4epkqL3upuVGewZJBaZ3rfIIzUvOsHMrnmjyeanoTQD2phF2ltvDbb3xxMMzAKcvvdUkhKI=
```

### CLI Share Combining

You can use the CLI to enter shares in order to recover a secret. Of course
you will need at least the number of shares necessary as determined
by the threshold when your shares were created. The `threshold` is visible
as the third field in every `HUMAN` formatted share.

As with splitting a secret, there are also three methods of getting the shares
into the CLI. `STDIN`, a path to a file containing shares, or interactively.

Here are some simple examples of using each:

**Example : `STDIN`**

```text
$ echo 'a secret' | tss split | tss combine

RECOVERED SECRET METADATA
*************************
hash : d4ea4551e9ff2cf56303875b1901fb8608a6164260c3b20c0976c7b606a4efc0
hash_alg : SHA256
identifier : fff58c38b14734f3
process_time : 0.34ms
threshold : 3
secret :
a secret
```

**Example : `--input-file`**

```text
$ echo 'a secret' | tss split -O /tmp/shares.txt
$ tss combine -I /tmp/shares.txt

RECOVERED SECRET METADATA
*************************
hash : d4ea4551e9ff2cf56303875b1901fb8608a6164260c3b20c0976c7b606a4efc0
hash_alg : SHA256
identifier : ae2983e30e0471fe
process_time : 0.47ms
threshold : 3
secret :
a secret
```

**Example : `interactive`**

```text
$ cat /tmp/shares.txt
# THRESHOLD SECRET SHARING SHARES
# 2017-01-29T02:01:00Z
# https://github.com/grempe/tss-rb


tss~v1~b9f7f87bc83fd89b~3~YjlmN2Y4N2JjODNmZDg5YgIDACoBbga2N__A9QOleLhC5R5b7stJ26iMPNpQ95YpmK-tWIrd-CYcrXGmcys=
tss~v1~b9f7f87bc83fd89b~3~YjlmN2Y4N2JjODNmZDg5YgIDACoCGXsZkLOG7uDOVN1Zn46pqftX4noA8LfXugg14KBfSggYP9fw4Ce10YA=
tss~v1~b9f7f87bc83fd89b~3~YjlmN2Y4N2JjODNmZDg5YgIDACoDFl3cwi80fpdh-I9eK3kNa8V9OlXX1Wx8y5a6bk2S0TDJzocr-1C3TWs=
tss~v1~b9f7f87bc83fd89b~3~YjlmN2Y4N2JjODNmZDg5YgIDACoEEspmyzmbnrSr7v41FFlHVTqQ6dx4aho8q5T1CBHQbNimdCv3gVXS50I=
tss~v1~b9f7f87bc83fd89b~3~YjlmN2Y4N2JjODNmZDg5YgIDACoFHeyjmaUpDsMEQqwyoK7jlwS6MfOvT8GX2gp6hvwd9-B3hXssmiLQe6k=

$ tss combine
Enter shares, one per line, and a dot (.) on a line by itself to finish :
share>  tss~v1~b9f7f87bc83fd89b~3~YjlmN2Y4N2JjODNmZDg5YgIDACoBbga2N__A9QOleLhC5R5b7stJ26iMPNpQ95YpmK-tWIrd-CYcrXGmcys=
share>  tss~v1~b9f7f87bc83fd89b~3~YjlmN2Y4N2JjODNmZDg5YgIDACoCGXsZkLOG7uDOVN1Zn46pqftX4noA8LfXugg14KBfSggYP9fw4Ce10YA=
share>  tss~v1~b9f7f87bc83fd89b~3~YjlmN2Y4N2JjODNmZDg5YgIDACoDFl3cwi80fpdh-I9eK3kNa8V9OlXX1Wx8y5a6bk2S0TDJzocr-1C3TWs=
share>  .

RECOVERED SECRET METADATA

CHANGELOG.md
# CHANGELOG

## v0.5.0 (1/28/2017)

* Breaking refactor to use automatic PKCS#7 padding on secrets w/ 16 byte block size.
* Update Copyright year
* Ruby 2.4.0 testing
* Update version of sysrandom gem
* Fixed minitest spec warning

## v0.4.2 (10/12/2016)

* Sign the gem with a new cert that expires in 10 years.
    Both old and new public certs are in the certs dir.

## v0.4.1 (9/28/2016)

* Use activesupport for blank support. Remove the extraction.
* Update sysrandom and remove the workaround needed for earlier version

## v0.4.0 (9/24/2016)

* Breaking change to force upcasing of some addition string args
* Use yard-contracts
* yardoc cleanups
* Remove int_commas utility method
* Hash w/ sha256 a,b strings in secure_compare
* Deeper Contracts integration

## v0.3.0 (9/24/2016)

* Breaking change, identifier cannot be an empty string
* Much greater coverage of functions with Contracts
* Related documentation updates

## v0.2.0 (9/23/2016)

* clone the shares object passed to TSS.combine to prevent modification
* test enhancements
* use cryptosphere/sysrandom in place of native securerandom
* integer args are no longer coercible from Strings
* Remove dry-* in favor of contracts (http://egonschiele.github.io/contracts.ruby/)
* readme fixes

## v0.1.1 (4/14/2016)

* documentation enhancements
* added two additional custom exception classes to allow rescuing from no secret recovery or invalid hash during recovery
* specify Rubies >= 2.1.0 in gemspec

## v0.1.0 (4/12/2016)

This is the initial ALPHA quality release of the tss gem.

It is for review only and should not yet be used in production.


lib/tss/util.rb
module TSS
  # Common utility, math, and conversion functions.
  module Util
    include Contracts::Core
    C = Contracts

    # The regex to match against human style shares
    HUMAN_SHARE_RE = /^tss~v1~*[a-zA-Z0-9\.\-\_]{0,16}~[0-9]{1,3}~([a-zA-Z0-9\-\_]+\={0,2})$/

    # The EXP table.  The elements are to be read from top to
    # bottom and left to right.  For example, EXP[0] is 0x01, EXP[8] is
    # 0x1a, and so on. Note that the EXP[255] entry is present only as a
    # placeholder, and is not actually used in any computation.
    EXP = [0x01, 0x03, 0x05, 0x0f, 0x11, 0x33, 0x55, 0xff,
         0x1a, 0x2e, 0x72, 0x96, 0xa1, 0xf8, 0x13, 0x35,
         0x5f, 0xe1, 0x38, 0x48, 0xd8, 0x73, 0x95, 0xa4,
         0xf7, 0x02, 0x06, 0x0a, 0x1e, 0x22, 0x66, 0xaa,
         0xe5, 0x34, 0x5c, 0xe4, 0x37, 0x59, 0xeb, 0x26,
         0x6a, 0xbe, 0xd9, 0x70, 0x90, 0xab, 0xe6, 0x31,
         0x53, 0xf5, 0x04, 0x0c, 0x14, 0x3c, 0x44, 0xcc,
         0x4f, 0xd1, 0x68, 0xb8, 0xd3, 0x6e, 0xb2, 0xcd,
         0x4c, 0xd4, 0x67, 0xa9, 0xe0, 0x3b, 0x4d, 0xd7,
         0x62, 0xa6, 0xf1, 0x08, 0x18, 0x28, 0x78, 0x88,
         0x83, 0x9e, 0xb9, 0xd0, 0x6b, 0xbd, 0xdc, 0x7f,
         0x81, 0x98, 0xb3, 0xce, 0x49, 0xdb, 0x76, 0x9a,
         0xb5, 0xc4, 0x57, 0xf9, 0x10, 0x30, 0x50, 0xf0,
         0x0b, 0x1d, 0x27, 0x69, 0xbb, 0xd6, 0x61, 0xa3,
         0xfe, 0x19, 0x2b, 0x7d, 0x87, 0x92, 0xad, 0xec,
         0x2f, 0x71, 0x93, 0xae, 0xe9, 0x20, 0x60, 0xa0,
         0xfb, 0x16, 0x3a, 0x4e, 0xd2, 0x6d, 0xb7, 0xc2,
         0x5d, 0xe7, 0x32, 0x56, 0xfa, 0x15, 0x3f, 0x41,
         0xc3, 0x5e, 0xe2, 0x3d, 0x47, 0xc9, 0x40, 0xc0,
         0x5b, 0xed, 0x2c, 0x74, 0x9c, 0xbf, 0xda, 0x75,
         0x9f, 0xba, 0xd5, 0x64, 0xac, 0xef, 0x2a, 0x7e,
         0x82, 0x9d, 0xbc, 0xdf, 0x7a, 0x8e, 0x89, 0x80,
         0x9b, 0xb6, 0xc1, 0x58, 0xe8, 0x23, 0x65, 0xaf,
         0xea, 0x25, 0x6f, 0xb1, 0xc8, 0x43, 0xc5, 0x54,
         0xfc, 0x1f, 0x21, 0x63, 0xa5, 0xf4, 0x07, 0x09,
         0x1b, 0x2d, 0x77, 0x99, 0xb0, 0xcb, 0x46, 0xca,
         0x45, 0xcf, 0x4a, 0xde, 0x79, 0x8b, 0x86, 0x91,
         0xa8, 0xe3, 0x3e, 0x42, 0xc6, 0x51, 0xf3, 0x0e,
         0x12, 0x36, 0x5a, 0xee, 0x29, 0x7b, 0x8d, 0x8c,
         0x8f, 0x8a, 0x85, 0x94, 0xa7, 0xf2, 0x0d, 0x17,
         0x39, 0x4b, 0xdd, 0x7c, 0x84, 0x97, 0xa2, 0xfd,
         0x1c, 0x24, 0x6c, 0xb4, 0xc7, 0x52, 0xf6, 0x00].freeze

    # The LOG table. The elements are to be read from top to
    # bottom and left to right. For example, LOG[1] is 0, LOG[8] is 75,
    # and so on. Note that the LOG[0] entry is present only as a
    # placeholder, and is not actually used in any computation.
    LOG = [0,     0,   25,    1,   50,    2,   26,  198,
         75,  199,   27,  104,   51,  238,  223,    3,
         100,   4,  224,   14,   52,  141,  129,  239,
         76,  113,    8,  200,  248,  105,   28,  193,
         125, 194,   29,  181,  249,  185,   39,  106,
         77,  228,  166,  114,  154,  201,    9,  120,
         101,  47,  138,    5,   33,   15,  225,   36,
         18,  240,  130,   69,   53,  147,  218,  142,
         150, 143,  219,  189,   54,  208,  206,  148,
         19,   92,  210,  241,   64,   70,  131,   56,
         102, 221,  253,   48,  191,    6,  139,   98,
         179,  37,  226,  152,   34,  136,  145,   16,
         126, 110,   72,  195,  163,  182,   30,   66,
         58,  107,   40,   84,  250,  133,   61,  186,
         43,  121,   10,   21,  155,  159,   94,  202,
         78,  212,  172,  229,  243,  115,  167,   87,
         175,  88,  168,   80,  244,  234,  214,  116,
         79,  174,  233,  213,  231,  230,  173,  232,
         44,  215,  117,  122,  235,   22,   11,  245,
         89,  203,   95,  176,  156,  169,   81,  160,
         127,  12,  246,  111,   23,  196,   73,  236,
         216,  67,   31,   45,  164,  118,  123,  183,
         204, 187,   62,   90,  251,   96,  177,  134,
         59,   82,  161,  108,  170,   85,   41,  157,
         151, 178,  135,  144,   97,  190,  220,  252,
         188, 149,  207,  205,   55,   63,   91,  209,
         83,   57,  132,   60,   65,  162,  109,   71,
         20,   42,  158,   93,   86,  242,  211,  171,
         68,   17,  146,  217,   35,   32,   46,  137,
         180, 124,  184,   38,  119,  153,  227,  165,
         103,  74,  237,  222,  197,   49,  254,   24,
         13,   99,  140,  128,  192,  247,  112,    7].freeze

    # GF(256) Addition
    # The addition operation returns the Bitwise
    # Exclusive OR (XOR) of its operands.
    #
    # @param a [Integer] a single Integer
    # @param b [Integer] a single Integer
    # @return [Integer] a GF(256) SUM of a and b
    def self.gf256_add(a, b)
      a ^ b
    end

    # The subtraction operation is identical to GF(256) addition, because the
    # field has characteristic two.
    #
    # @param a [Integer] a single Integer
    # @param b [Integer] a single Integer
    # @return [Integer] a GF(256) subtraction of a and b
    def self.gf256_sub(a, b)
      gf256_add(a, b)
    end

    # The multiplication operation takes two elements X and Y as input and
    # proceeds as follows.  If either X or Y is equal to 0x00, then the
    # operation returns 0x00.  Otherwise, the value EXP[ (LOG[X] + LOG[Y])
    # modulo 255] is returned.
    #
    # @param x [Integer] a single Integer
    # @param y [Integer] a single Integer
    # @return [Integer] a GF(256) multiplication of x and y
    def self.gf256_mul(x, y)
      return 0 if x == 0 || y == 0
      EXP[(LOG[x] + LOG[y]) % 255]
    end

    # The division operation takes a dividend X and a divisor Y as input
    # and computes X divided by Y as follows.  If X is equal to 0x00, then
    # the operation returns 0x00.  If Y is equal to 0x00, then the input is
    # invalid, and an error condition occurs.  Otherwise, the value
    # EXP[(LOG[X] - LOG[Y]) modulo 255] is returned.
    #
    # @param x [Integer] a single Integer
    # @param y [Integer] a single Integer
    # @return [Integer] a GF(256) division of x divided by y
    # @raise [TSS::Error] if an attempt to divide by zero is tried
    def self.gf256_div(x, y)
      return 0 if x == 0
      raise TSS::Error, 'divide by zero' if y == 0
      EXP[(LOG[x] - LOG[y]) % 255]
    end

    # Share generation Function
    #
    # The function f takes as input a single octet X that is not equal to
    # 0x00, and an array A of M octets, and returns a single octet.  It is
    # defined as:
    #
    #   f(X, A) =  GF_SUM A[i] (*) X^i
    #              i=0,M-1
    #
    # Because the GF_SUM summation takes place over GF(256), each addition
    # uses the exclusive-or operation, and not integer addition.  Note that
    # the successive values of X^i used in the computation of the function
    # f can be computed by multiplying a value by X once for each term in
    # the summation.
    #
    # @param x [Integer] a single Integer
    # @param bytes [Array<Integer>] an Array of Integers
    # @return [Integer] a single Integer
    # @raise [TSS::Error] if the index value for the share is zero
    def self.f(x, bytes)
      raise TSS::Error, 'invalid share index value, cannot be 0' if x == 0
      y = 0
      x_i = 1

      bytes.each do |b|
        y = gf256_add(y, gf256_mul(b, x_i))
        x_i = gf256_mul(x_i, x)
      end

      y
    end

    # Secret Reconstruction Function
    #
    # We define the function L_i (for i from 0 to M-1, inclusive) that
    # takes as input an array U of M pairwise distinct octets, and is
    # defined as
    #
    #                             U[j]
    #   L_i(U) = GF_PRODUCT   -------------
    #            j=0,M-1, j!=i  U[j] (+) U[i]
    #
    # Here the product runs over all of the values of j from 0 to M-1,
    # excluding the value i.  (This function is equal to ith Lagrange
    # function, evaluated at zero.)  Note that the denominator in the above
    # expression is never equal to zero because U[i] is not equal to U[j]
    # whenever i is not equal to j.
    #
    # @param i [Integer] a single Integer
    # @param u [Array<Integer>] an Array of Integers
    # @return [Integer] a single Integer
    def self.basis_poly(i, u)
      prod = 1

      (0..(u.length - 1)).each do |j|
        next if i == j
        prod = gf256_mul(prod, gf256_div(u[j], gf256_add(u[j], u[i])))
      end

      prod
    end

    # Secret Reconstruction Function
    #
    # We denote the interpolation function as I. This function takes as
    # input two arrays U and V, each consisting of M octets, and returns a
    # single octet; it is defined as:
    #
    #   I(U, V) =  GF_SUM  L_i(U) (*) V[i].
    #              i=0,M-1
    #
    # @param u [Array<Integer>] an Array of Integers
    # @param v [Array<Integer>] an Array of Integers
    # @return [Integer] a single Integer
    def self.lagrange_interpolation(u, v)
      sum = 0

      (0..(v.length - 1)).each do |i|
        sum = gf256_add(sum, gf256_mul(basis_poly(i, u), v[i]))
      end

      sum
    end

    # Convert a UTF-8 String to an Array of Bytes
    #
    # @param str a UTF-8 String to convert
    # @return an Array of Integer Bytes
    Contract String => C::ArrayOf[C::Int]
    def self.utf8_to_bytes(str)
      str.bytes.to_a
    end

    # Convert an Array of Bytes to a UTF-8 String
    #
    # @param bytes an Array of Bytes to convert
    # @return a UTF-8 String
    Contract C::ArrayOf[C::Int] => String
    def self.bytes_to_utf8(bytes)
      bytes.pack('C*').force_encoding('utf-8')
    end

    # Convert an Array of Bytes to a hex String
    #
    # @param bytes an Array of Bytes to convert
    # @return a hex String
    Contract C::ArrayOf[C::Int] => String
    def self.bytes_to_hex(bytes)
      hex = ''
      bytes.each { |b| hex += sprintf('%02x', b) }
      hex.downcase
    end

    # Convert a hex String to an Array of Bytes
    #
    # @param str a hex String to convert
    # @return an Array of Integer Bytes
    Contract String => C::ArrayOf[C::Int]
    def self.hex_to_bytes(str)
      [str].pack('H*').unpack('C*')
    end

    # Convert a hex String to a UTF-8 String
    #
    # @param hex a hex String to convert
    # @return a UTF-8 String
    Contract String => String
    def self.hex_to_utf8(hex)
      bytes_to_utf8(hex_to_bytes(hex))
    end

    # Convert a UTF-8 String to a hex String
    #
    # @param str a UTF-8 String to convert
    # @return a hex String
    Contract String => String
    def self.utf8_to_hex(str)
      bytes_to_hex(utf8_to_bytes(str))
    end

    # Pad a String with PKCS#7 (RFC5652)
    # See : https://tools.ietf.org/html/rfc5652#section-6.3
    #
    # @param str the String or Array of bytes to pad
    # @param k pad blocksize (0-255), default 16
    # @return a PKCS#7 padded String or Array of bytes
    Contract C::Or[Array, String], C::PadBlocksizeArg => C::Or[Array, String]
    def self.pad(str, k = TSS::PADDING_BLOCK_SIZE_BYTES)
      return str if k.zero?
      str_bytes = str.is_a?(Array) ? str : TSS::Util.utf8_to_bytes(str)
      l = str_bytes.length
      val = k - (l % k)
      pad_bytes = [val] * val
      padded_str_bytes = str_bytes + pad_bytes
      str.is_a?(Array) ? padded_str_bytes : TSS::Util.bytes_to_utf8(padded_str_bytes)
    end

    # Remove padding from a String previously padded with PKCS#7 (RFC5652)
    #
    # @param str the String to remove padding from
    # @param k pad blocksize (0-255)
    # @return an unpadded String or Array of bytes
    Contract C::Or[Array, String], C::PadBlocksizeArg => C::Or[Array, String]
    def self.unpad(str, k = TSS::PADDING_BLOCK_SIZE_BYTES)
      return str if k.zero?
      str_bytes = str.is_a?(Array) ? str : TSS::Util.utf8_to_bytes(str)
      val = str_bytes.last
      raise 'Input is not padded or padding is corrupt' if val > k
      # Verify that the proper number of PKCS#7 padding bytes are present
      # and match the last byte value in both value and number of bytes present.
      raise 'Padding bytes are invalid' unless str_bytes.last(val).all? {|b| b == val}
      l = str_bytes.length - val
      unpadded_str_bytes = str_bytes.take(l)
      str.is_a?(Array) ? unpadded_str_bytes : TSS::Util.bytes_to_utf8(unpadded_str_bytes)
    end

    # Constant time string comparison.
    #   Extracted from Rack::Utils
    #   https://github.com/rack/rack/blob/master/lib/rack/utils.rb
    #
    #   NOTE: the values compared should be of fixed length, such as strings
    #   that have already been hashed. This should not be used
    #   on variable length plaintext strings because it could leak length info
    #   via timing attacks.
    #
    #   The user provided value should always be passed in as the second
    #   parameter so as not to leak info about the secret.
    #
    # @param a the private value
    # @param b the user provided value
    # @return whether the strings match or not
    Contract String, String => C::Bool
    def self.secure_compare(a, b)
      return false unless a.bytesize == b.bytesize
      ah = Digest::SHA256.hexdigest(a)
      bh = Digest::SHA256.hexdigest(b)

      l = ah.unpack('C*')
      r, i = 0, -1
      bh.each_byte { |v| r |= v ^ l[i+=1] }
      r == 0
    end

    # Extract the header data from a binary share.
    # Extra "\x00" padding in the identifier will be removed.
    #
    # @param share a binary octet share
    # @return header attributes
    Contract String => Hash
    def self.extract_share_header(share)
      h = Splitter::SHARE_HEADER_STRUCT.decode(share)
      h[:identifier] = h[:identifier].delete("\x00")
      return h
    end

    # Calculate the factorial for an Integer.
    #
    # @param n the Integer to calculate for
    # @return the factorial of n
    Contract C::Int => C::Int
    def self.factorial(n)
      (1..n).reduce(:*) || 1
    end

    # Calculate the number of combinations possible
    # for a given number of shares and threshold.
    #
    # * http://www.wolframalpha.com/input/?i=20+choose+5
    # * http://www.mathsisfun.com/combinatorics/combinations-permutations-calculator.html (Set balls, 20, 5, no, no) == 15504
    # * http://www.mathsisfun.com/combinatorics/combinations-permutations.html
    # * https://jdanger.com/calculating-factorials-in-ruby.html
    # * http://chriscontinanza.com/2010/10/29/Array.html
    # * http://stackoverflow.com/questions/2434503/ruby-factorial-function
    #
    # @param n the total number of shares
    # @param r the threshold number of shares
    # @return the number of possible combinations
    Contract C::Int, C::Int => C::Int
    def self.calc_combinations(n, r)
      factorial(n) / (factorial(r) * factorial(n - r))
    end
  end
end
